src/python/pants/option/parser.py

```diff
@@ -581,28 +581,40 @@ def expand(val_or_str):
             else:
                 return val_or_str
 
+        # Helper function to merge multiple values from a single rank (e.g., multiple flags,
+        # or multiple config files).
+        def merge_in_rank(vals):
+            if not vals:
+                return None
+            expanded_vals = [to_value_type(expand(x)) for x in vals]
+            if is_list_option(kwargs):
+                return ListValueComponent.merge(expanded_vals)
+            if is_dict_option(kwargs):
+                return DictValueComponent.merge(expanded_vals)
+            return expanded_vals[-1]  # Last value wins.
+
         # Get value from config files, and capture details about its derivation.
         config_details = None
         config_section = GLOBAL_SCOPE_CONFIG_SECTION if self._scope == GLOBAL_SCOPE else self._scope
-        config_default_val_or_str = expand(self._config.get(DEFAULT_SECTION, dest))
-        config_val_or_str = expand(self._config.get(config_section, dest))
+        config_default_val = merge_in_rank(self._config.get(DEFAULT_SECTION, dest))
```

Comment:
Previously we pre-converted flag values, but left config and env var values as strings-or-values (usually strings, unless decoded from a json/yaml fromfile), to be converted later.

Now we have to also pre-convert the config file values, so we can handle list values properly. So we may as well also do so for the env var value, for uniformity.